File: nomdl/codegen/test/struct_optional_test.go
package test

import (
	"testing"

	"github.com/attic-labs/noms/Godeps/_workspace/src/github.com/stretchr/testify/assert"
	"github.com/attic-labs/noms/chunks"
	"github.com/attic-labs/noms/nomdl/codegen/test/gen"
)

func TestOptionalStruct(t *testing.T) {
	assert := assert.New(t)
	cs := chunks.NewMemoryStore()

	str := gen.NewOptionalStruct(cs)

	_, ok := str.S()
	assert.False(ok)

	_, ok = str.B()
	assert.False(ok)

	str = str.SetS("hi")
	s, ok := str.S()
	assert.True(ok)
	assert.Equal("hi", s)

	str = str.SetB(false)
	b, ok := str.B()
	assert.True(ok)
	assert.False(b)
}

func TestOptionalStructDef(t *testing.T) {
	assert := assert.New(t)
	cs := chunks.NewMemoryStore()

	def := gen.OptionalStructDef{}
	str := def.New(cs)
	s, ok := str.S()
	assert.True(ok)
	assert.Equal("", s)

	b, ok := str.B()
	assert.True(ok)
	assert.False(b)

	def2 := str.Def()
	assert.Equal(def, def2)
}

func TestOptionalStructDefFromNew(t *testing.T) {
	assert := assert.New(t)
	cs := chunks.NewMemoryStore()

	str := gen.NewOptionalStruct(cs).SetB(true)
	def := str.Def()
	def2 := gen.OptionalStructDef{B: true}
	assert.Equal(def, def2)
}

File: nomdl/codegen/test/struct-optional-test.js
// @flow

import {assert} from 'chai';
import {suite, test} from 'mocha';

import {OptionalStruct} from './gen/struct_optional.noms.js';

suite('struct_optional.noms', () => {
  test('constructor', async () => {
    const os = new OptionalStruct({});
    assert.isUndefined(os.s);
    assert.isUndefined(os.b);

    const os2 = os.setS('hi');
    assert.equal(os2.s, 'hi');
    assert.isUndefined(os2.b);

    const os3 = os2.setB(true);
    assert.equal(os3.s, 'hi');
    assert.equal(os3.b, true);

    const os4 = os2.setB(undefined).setS(undefined);
    assert.isTrue(os4.equals(os));
  });
});
